test/aleph/resource_leak_detector.clj

```diff
@@ -0,0 +1,210 @@
+  (when-not (= ResourceLeakDetector$Level/PARANOID (ResourceLeakDetector/getLevel))
+    (throw (RuntimeException.
+            (str "`aleph.resource_leak_detector` requires `-Dio.netty.leakDetection.level=PARANOID`. "
+                 "Current level is `" (ResourceLeakDetector/getLevel) "`.")))))
+
+(def +max-probe-gc-runs+
+  "Maximum number of times the GC will be run to detect a leaked probe."
+  10)
+
+(def +probe-hint-marker+
+  "ALEPH LEAK DETECTOR PROBE")
+
+(defn hint-record-pattern [hint-pattern]
+  (re-pattern (str "(?m)^\\s*Hint: " hint-pattern "$")))
+
+(def +probe-hint-pattern+
+  (hint-record-pattern (str +probe-hint-marker+ " \\d+")))
+
+(defn probe? [leak]
+  (re-find +probe-hint-pattern+ (:records leak)))
+
+(defn contains-hint? [hint leak]
+  (re-find (hint-record-pattern hint) (:records leak)))
+
+(defn remove-probes [leaks]
+  (remove probe? leaks))
+
+(let [cnt (atom 0)]
+  (defn gen-probe-hint []
+    (str +probe-hint-marker+ " " (swap! cnt inc))))
+
+(defn leak-probe! [hint]
+  (-> AbstractByteBufAllocator/DEFAULT
+      (.buffer 1)
+      (.touch hint)))
+
+(def current-leaks)
+
+(defn force-leak-detection! []
+  (System/gc)
+  (System/runFinalization)
+  ;; Transitively trigger a track() invocation which in turn works
+  ;; off the leaked references queue.
+  (-> AbstractByteBufAllocator/DEFAULT (.buffer 1) .release))
+
+(defn await-probe! [probe-hint]
+  (loop [n +max-probe-gc-runs+]
+    (force-leak-detection!)
+    (if (zero? n)
+      (throw (RuntimeException. "Gave up awaiting leak probe. Try increasing +max-probe-gc-runs+."))
+      (when-not (some (partial contains-hint? probe-hint) @current-leaks)
+        (recur (dec n))))))
+
+(defn with-leak-collection
+  "Invokes thunk `f` and tries hard to collect any resource leaks it may have caused.
+
+  It works as follows: After invoking `f`, it intentionally leaks a (small) buffer, marked as a
+  probe. It then runs the garbage collector and polls the leak detector in a loop until it reports a
+  leak which matches the probe. Eventually, it invokes `handle-leaks` with a sequence of any other
+  detected leaks it collected along the way (empty when none were detected).
+
+  A leak is represented as a map with the following keys:
+  - `:resource-type` is a string with the name of the leaked resource type (e.g. \"ByteBuf\")
+  - `:records` is a multi-line string which holds the trace of the leak.
+
+  Requires the leak detector to be `enabled?`.
+
+  When nested, each child establishes a fresh leak collection scope. However, this only works within
+  the same thread, so any asynchronous processes started by and outliving `f` will leak into the
+  parent scope(s)."
+  [f handle-leaks]
+  (ensure-consistent-config!)
+  (with-redefs [current-leaks (atom [])]
+    (f)
+    (let [hint (gen-probe-hint)]
+      (leak-probe! hint)
+      (await-probe! hint)
+      (handle-leaks (remove-probes @current-leaks)))))
```

Comment:
Interesting technique. Is there some doc on it? 

Why is a probe needed? Why not listen for all leaks when running `(f)`?

Is there a race? Can we miss leaks because we call `f` before setting up leak detection?

If we ever run parallel tests, using https://github.com/mourjo/dynamic-redef might help.

Reply:
> Interesting technique. Is there some doc on it?

Nothing beyond what's in the docstring of `with-leak-collection`.

> Why is a probe needed? Why not listen for all leaks when running `(f)`?

We do listen for all leaks when running `(f)` via the `(with-redefs [current-leaks (atom [])] ...)`.

To understand why the probe is needed, you need to know that the detector can only detect leaked resources after they have been marked as garbage by the GC. Now since the probe is only leaked *after* `f` has returned, there's a higher chance that any leaks which occurred during `f` will have been detected by the time the detector reports the probe again. There's still a small chance that the GC will mark the probe as garbage before it marked all resources leaked by `f`, which is why we end up with only ~95% accuracy. 

> Is there a race?

Not really a race but more of a gap of unreliability due to the GC being non-deterministic. Even our explicit `System/gc`  loop doesn't help since that only serves as a hint to the GC. I did extensive research on whether there is a way to force a full GC somehow, but it appears there isn't :confused: 

> Can we miss leaks because we call `f` before setting up leak detection?

As mentioned above, leak detection is indeed set up before `f`!


> If we ever run parallel tests, using https://github.com/mourjo/dynamic-redef might help.

Indeed, though I would tend to always run tests serially for leak detection since the GC is the most unpredictable component involved. Since the GC owuld be shared across all threads, I would expect this to further reduce reliability. But we can cross that bridge when we get to it I guess :sweat_smile:

Reply:
Sounds good. I think the word "probe" confused me, because, IIUC, the probe isn't "probing" for leaks itself, but is used as a marker to indicate which leaks were caused by `f` (i.e., all leaks before the probe). Is that right, or am I still misunderstanding?

Reply:
Yep, that's right! Or ... well, it isn't really a *marker* for which leaks came before it (during `f`) but rather as a *trigger* to flush out any leaks which happened during `f` but which have not yet been marked by the GC. So maybe `gc-trigger` could be a better name? Any other ideas?